### PR TITLE
[CodeGen][NewPM] Port SpillPlacement analysis to NPM

### PR DESCRIPTION
**Pull request by optimisan**
I am not sure how to test this.
Porting Greedy regalloc next.

### CODE REVIEW

**Comment by optimisan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116618?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#116618** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116618?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116618?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#116617** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116617?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116616** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/116616?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: llvm/lib/CodeGen/SpillPlacement.h [hunk 1]
```diff
@@ -38,13 +39,21 @@ class BitVector;
...
-class SpillPlacement : public MachineFunctionPass {
   struct Node;
+
   const MachineFunction *MF = nullptr;
   const EdgeBundles *bundles = nullptr;
   const MachineBlockFrequencyInfo *MBFI = nullptr;
-  Node *nodes = nullptr;
+
+  static void arrayDeleter(Node *N);
+  std::unique_ptr<Node, decltype(&arrayDeleter)> nodes;
```
Comment (paperchalice) [inline llvm/lib/CodeGen/SpillPlacement.h:56]: Array form would be better:
```suggestion
  std::unique_ptr<Node[]> nodes;
```
https://en.cppreference.com/w/cpp/memory/unique_ptr

File: llvm/lib/CodeGen/SpillPlacement.cpp [hunk 2]
```diff
@@ -189,32 +189,57 @@ struct SpillPlacement::Node {
...
+  SpillPlacement Impl(Bundles, MBFI);
+  Impl.run(MF);
+  return Impl;
+}
+
+bool SpillPlacementAnalysis::Result::invalidate(
+    MachineFunction &MF, const PreservedAnalyses &PA,
+    MachineFunctionAnalysisManager::Invalidator &Inv) {
+  auto PAC = PA.getChecker<SpillPlacementAnalysis>();
+  return !(PAC.preserved() ||
```
Comment (arsenm) [inline llvm/lib/CodeGen/SpillPlacement.cpp:217]: demorgan condition

File: llvm/lib/CodeGen/SpillPlacement.cpp [hunk 3]
```diff
@@ -189,32 +189,57 @@ struct SpillPlacement::Node {
   }
 };
 
-bool SpillPlacement::runOnMachineFunction(MachineFunction &mf) {
+bool SpillPlacementWrapperLegacy::runOnMachineFunction(MachineFunction &MF) {
+  auto *Bundles = &getAnalysis<EdgeBundlesWrapperLegacy>().getEdgeBundles();
+  auto *MBFI = &getAnalysis<MachineBlockFrequencyInfoWrapperPass>().getMBFI();
+
+  Impl.reset(new SpillPlacement(Bundles, MBFI));
```
Comment (arsenm) [inline llvm/lib/CodeGen/SpillPlacement.cpp:196]: Don't see why this needs to be heap allocated and not just a pass member

File: llvm/lib/CodeGen/SpillPlacement.h (hunk 1 above)
Comment (optimisan) [reply llvm/lib/CodeGen/SpillPlacement.h:56]: The definition of `Node` is not available here, so the default deleter fails to compile sizeof(Node) for this incomplete type. To hack around it I put the definition of `arrayDeleter` in the implementation where struct Node is defined.

But changing to `unique_ptr<Node[], theDeleter>` facilitates removal of `.get()` calls

Comment (paperchalice) [reply llvm/lib/CodeGen/SpillPlacement.h]: An outlined default destructor would work. 🤔

Comment (optimisan) [reply llvm/lib/CodeGen/SpillPlacement.h]: That works

File: llvm/include/llvm/Passes/MachinePassRegistry.def [hunk 4]
```diff
@@ -112,6 +112,7 @@ MACHINE_FUNCTION_ANALYSIS("machine-post-dom-tree",
                           MachinePostDominatorTreeAnalysis())
 MACHINE_FUNCTION_ANALYSIS("machine-trace-metrics", MachineTraceMetricsAnalysis())
 MACHINE_FUNCTION_ANALYSIS("pass-instrumentation", PassInstrumentationAnalysis(PIC))
+MACHINE_FUNCTION_ANALYSIS("spill-code-placement", SpillPlacementAnalysis())
```
Comment (arsenm) [inline llvm/include/llvm/Passes/MachinePassRegistry.def:115]: Alphabetical order

File: llvm/lib/CodeGen/SpillPlacement.cpp [hunk 5]
```diff
@@ -189,32 +189,63 @@ struct SpillPlacement::Node {
...
+  SpillPlacement Impl;
+  Impl.run(MF, Bundles, MBFI);
+  return Impl;
+}
+
+bool SpillPlacementAnalysis::Result::invalidate(
+    MachineFunction &MF, const PreservedAnalyses &PA,
+    MachineFunctionAnalysisManager::Invalidator &Inv) {
+  auto PAC = PA.getChecker<SpillPlacementAnalysis>();
+  return (!PAC.preserved() &&
```
Comment (arsenm) [inline llvm/lib/CodeGen/SpillPlacement.cpp:216]: demorgan this

File: llvm/include/llvm/CodeGen/SpillPlacement.h [hunk 6]
```diff
@@ -154,17 +157,52 @@ class SpillPlacement : public MachineFunctionPass {
     return BlockFrequencies[Number];
   }
 
+  bool invalidate(MachineFunction &MF, const PreservedAnalyses &PA,
+                  MachineFunctionAnalysisManager::Invalidator &Inv);
+
+  // Move the default definitions to the implementation
+  // so the full definition of Node is available.
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/SpillPlacement.h:164]: Not sure what this comment is supposed to mean

File: llvm/include/llvm/CodeGen/SpillPlacement.h [hunk 7]
```diff
@@ -154,17 +157,51 @@ class SpillPlacement : public MachineFunctionPass {
     return BlockFrequencies[Number];
   }
 
+  bool invalidate(MachineFunction &MF, const PreservedAnalyses &PA,
+                  MachineFunctionAnalysisManager::Invalidator &Inv);
+
+  // Out of line definitions so unique_ptr<Node[]> doesn't need to be complete.
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/SpillPlacement.h:163]: Don't really need to comment this

**Review by arsenm — APPROVED** [review]

**Pull request merged**